pkg/filesystem/subvolume.go

```diff
@@ -0,0 +1,174 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subvolume
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/rook/kubectl-rook-ceph/pkg/exec"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type fsStruct struct {
+	Name string
+	data []string
+}
+
+type subVolumeInfo struct {
+	svg string
+	fs  string
+}
+
+func List(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStaleOnly bool) {
+
+	subvolumeNames := getK8sRefSubvolume(ctx, clientsets)
+	listCephFSSubvolumes(ctx, clientsets, operatorNamespace, clusterNamespace, includeStaleOnly, subvolumeNames)
+}
+
+// getk8sRefSubvolume returns the k8s ref for the subvolumes
+func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map[string]subVolumeInfo {
+	pvList, err := clientsets.Kube.CoreV1().PersistentVolumes().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching PVs: %v\n", err))
+	}
+	subvolumeNames := make(map[string]subVolumeInfo)
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI.VolumeAttributes["subvolumeName"] != "" {
+			subvolumeNames[pv.Spec.CSI.VolumeAttributes["subvolumeName"]] = subVolumeInfo{}
+		}
+	}
+	return subvolumeNames
+}
+
+// listCephFSSubvolumes list all the subvolumes
+func listCephFSSubvolumes(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStaleOnly bool, subvolumeNames map[string]subVolumeInfo) {
+
+	// getFilesystem gets the filesystem
+	fsstruct := getFileSystem(ctx, clientsets, operatorNamespace, clusterNamespace)
+	var svList string
+	fmt.Println("Filesystem  Subvolume  SubvolumeGroup  State")
+
+	// this iterates over the filesystems and subvolumegroup to get the list of subvolumes that exist
+	for _, fs := range fsstruct {
+		// gets the subvolumegroup in the filesystem
+		subvolg := getSubvolumeGroup(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name)
+		for _, svg := range subvolg {
+			svList = exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "ls", fs.Name, svg.Name}, operatorNamespace, clusterNamespace, true, false)
+			subvol := unMarshaljson(svList)
+			if len(subvol) == 0 {
+				continue
+			}
+			// append the subvolume which doesn't have any snapshot attached to it.
+			for _, sv := range subvol {
+				// Assume the volume is stale unless proven otherwise
+				stale := true
+				// lookup for subvolume in list of the PV references
+				_, ok := subvolumeNames[sv.Name]
+				if ok || checkSnapshot(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name, sv.Name, svg.Name) {
+					// The volume is not stale if a PV was found, or it has a snapshot
+					stale = false
```

Comment:
we can have a stale subvolume with a snapshot as well. we cannot consider if a subvolume has a snapshot as not a stale.

Reply:
But deleting the subvolume if it has a snapshot can lead to data loss, hence the plan is to let them know that the subvolume has snapshot so it is good to delete the snapshot manually and then re-try.
What do you think should be an ideal case?

Reply:
you can delete the subvolume with `–retain-snapshots` flag, this is take care of cleanup when the snapshot is deleted. this is what we do in cephcsi as well

Reply:
sure will test and update.